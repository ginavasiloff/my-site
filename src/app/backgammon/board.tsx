export type BoardProps = {
  width: number
  height: number
  trimWidth: number
  boardColor: string
  boardTrim: string
}

export const Board = (boardProps: BoardProps) => {
  const { boardColor, boardTrim, width, height, trimWidth } = boardProps

  return (
    <svg
      viewBox={`0 0 ${width} ${height}`}
      xmlns="http://www.w3.org/2000/svg"
      data-testid={'backgammon-board'}
    >
      <g>
        <rect width={width} height={height} x="0" y="0" fill={boardTrim} />
        <rect
          width={width - trimWidth * 2}
          height={height - trimWidth * 2}
          x={trimWidth}
          y={trimWidth}
          fill={boardColor}
        />
        <rect
          width={trimWidth * 2}
          height={height}
          x={width / 2 - trimWidth}
          y="0"
          fill={boardTrim}
        />
      </g>
      <Triangles {...boardProps} />
    </svg>
  )
}

const Triangles = (boardProps: BoardProps) => {
  const { boardTrim, width, height, trimWidth } = boardProps
  const dx = (width - trimWidth * 4) / 12
  const y0 = trimWidth
  const y2 = (height - trimWidth * 2) / 2 - height * 0.1
  const y3 = (height - trimWidth * 2) / 2 + height * 0.1
  const y4 = height - trimWidth

  let triangleCoordinates: { [key: string]: number[][] } = {}
  for (let i = 0; i < 6; i++) {
    const x0 = trimWidth + i * dx
    const x1 = x0 + dx
    const x2 = x0 + dx / 2
    const x3 = width / 2 + x0
    const x4 = x3 + dx
    const x5 = x4 - dx / 2

    triangleCoordinates = {
      ...triangleCoordinates,
      [i + 1]: [
        [x3, y4],
        [x4, y4],
        [x5, y3],
      ],
      [i + 7]: [
        [x0, y4],
        [x1, y4],
        [x2, y3],
      ],
      [i + 13]: [
        [x0, y0],
        [x1, y0],
        [x2, y2],
      ],
      [i + 19]: [
        [x3, y0],
        [x4, y0],
        [x5, y2],
      ],
    }
  }

  return Object.keys(triangleCoordinates).map((key, i) => {
    const points = triangleCoordinates[key]
    return <polygon key={i} points={points.join(', ')} fill={boardTrim} />
  })
}
