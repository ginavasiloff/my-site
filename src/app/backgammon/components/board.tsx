import { Coordinates, Positions } from '../types'
import { Pieces } from './pieces'

const boardColor = 'rgb(248, 195, 157)'
const boardTrim = 'rgb(139, 70, 36)'

type BoardDimensions = {
  width: number
  height: number
  trimWidth: number
  dx: number
}

const calculateTriangleCoordinates = ({
  width,
  height,
  trimWidth,
  dx,
}: BoardDimensions): Coordinates => {
  const y0 = trimWidth
  const y2 = (height - trimWidth * 2) / 2 - height * 0.1
  const y3 = (height - trimWidth * 2) / 2 + height * 0.1
  const y4 = height - trimWidth

  let triangleCoordinates = {}
  for (let i = 0; i < 6; i++) {
    const x0 = trimWidth + i * dx
    const x1 = x0 + dx
    const x2 = x0 + dx / 2
    const x3 = width / 2 + x0
    const x4 = x3 + dx
    const x5 = x4 - dx / 2

    triangleCoordinates = {
      ...triangleCoordinates,
      [6 - i]: [
        [x3, y4],
        [x4, y4],
        [x5, y3],
      ],
      [12 - i]: [
        [x0, y4],
        [x1, y4],
        [x2, y3],
      ],
      [i + 13]: [
        [x0, y0],
        [x1, y0],
        [x2, y2],
      ],
      [i + 19]: [
        [x3, y0],
        [x4, y0],
        [x5, y2],
      ],
    }
  }
  return triangleCoordinates
}

const Triangles = ({ coordinates }: { coordinates: Coordinates }) => {
  return Object.keys(coordinates).map((key, i) => {
    const points = coordinates[key]
    return (
      <polygon
        data-testid={i}
        key={i}
        points={points.join(', ')}
        fill={boardTrim}
      />
    )
  })
}

export const Board = ({ positions }: { positions: Positions }) => {
  const trimWidth = 1
  const width = 100
  const height = 100
  const dx = (width - trimWidth * 4) / 12

  const triangleCoordinates = calculateTriangleCoordinates({
    width,
    height,
    trimWidth,
    dx,
  })

  return (
    <svg
      viewBox={`0 0 ${width} ${height}`}
      xmlns="http://www.w3.org/2000/svg"
      data-testid={'backgammon-board'}
      width={width * 4}
      height={height * 4}
    >
      <g>
        <rect width={width} height={height} x="0" y="0" fill={boardTrim} />
        <rect
          width={width - trimWidth * 2}
          height={height - trimWidth * 2}
          x={trimWidth}
          y={trimWidth}
          fill={boardColor}
        />
        <rect
          width={trimWidth * 2}
          height={height}
          x={width / 2 - trimWidth}
          y="0"
          fill={boardTrim}
        />
      </g>
      <Triangles coordinates={triangleCoordinates} />
      <Pieces
        positions={positions}
        coordinates={triangleCoordinates}
        dx={dx}
        trimWidth={trimWidth}
      />
    </svg>
  )
}
