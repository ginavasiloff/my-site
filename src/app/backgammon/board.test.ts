// @vitest-environment jsdom

import { render } from '@testing-library/react'
import { Board } from './board'

describe('Board', () => {
  test('it renders', () => {
    const { getByTestId } = render(
      Board({
        trimWidth: 1,
        width: 100,
        height: 100,
        boardColor: 'white',
        boardTrim: 'black',
      })
    )
    expect(getByTestId('backgammon-board').children.length).toBeGreaterThan
    expect(getByTestId('backgammon-board')).toMatchSnapshot()
  })
})
