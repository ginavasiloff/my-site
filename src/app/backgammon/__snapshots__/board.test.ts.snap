// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Board > it renders 1`] = `
<svg
  data-testid="backgammon-board"
  viewBox="0 0 100 100"
  xmlns="http://www.w3.org/2000/svg"
>
  <g>
    <rect
      fill="black"
      height="100"
      width="100"
      x="0"
      y="0"
    />
    <rect
      fill="white"
      height="98"
      width="98"
      x="1"
      y="1"
    />
    <rect
      fill="black"
      height="100"
      width="2"
      x="49"
      y="0"
    />
  </g>
  <polygon
    fill="black"
    points="51,99, 59,99, 55,59"
  />
  <polygon
    fill="black"
    points="59,99, 67,99, 63,59"
  />
  <polygon
    fill="black"
    points="67,99, 75,99, 71,59"
  />
  <polygon
    fill="black"
    points="75,99, 83,99, 79,59"
  />
  <polygon
    fill="black"
    points="83,99, 91,99, 87,59"
  />
  <polygon
    fill="black"
    points="91,99, 99,99, 95,59"
  />
  <polygon
    fill="black"
    points="1,99, 9,99, 5,59"
  />
  <polygon
    fill="black"
    points="9,99, 17,99, 13,59"
  />
  <polygon
    fill="black"
    points="17,99, 25,99, 21,59"
  />
  <polygon
    fill="black"
    points="25,99, 33,99, 29,59"
  />
  <polygon
    fill="black"
    points="33,99, 41,99, 37,59"
  />
  <polygon
    fill="black"
    points="41,99, 49,99, 45,59"
  />
  <polygon
    fill="black"
    points="1,1, 9,1, 5,39"
  />
  <polygon
    fill="black"
    points="9,1, 17,1, 13,39"
  />
  <polygon
    fill="black"
    points="17,1, 25,1, 21,39"
  />
  <polygon
    fill="black"
    points="25,1, 33,1, 29,39"
  />
  <polygon
    fill="black"
    points="33,1, 41,1, 37,39"
  />
  <polygon
    fill="black"
    points="41,1, 49,1, 45,39"
  />
  <polygon
    fill="black"
    points="51,1, 59,1, 55,39"
  />
  <polygon
    fill="black"
    points="59,1, 67,1, 63,39"
  />
  <polygon
    fill="black"
    points="67,1, 75,1, 71,39"
  />
  <polygon
    fill="black"
    points="75,1, 83,1, 79,39"
  />
  <polygon
    fill="black"
    points="83,1, 91,1, 87,39"
  />
  <polygon
    fill="black"
    points="91,1, 99,1, 95,39"
  />
</svg>
`;
